### ISSUE TITLE
Object not assignable to type even after narrowing

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#39256]
Originally posted on [StackOverflow](https://stackoverflow.com/questions/62576620/typescript-object-not-assignable-to-type-even-after-narrowing). Re-posting here as I hope this is something which can be improved in TypeScript, so we don't have to resort to type assertions.

**TypeScript Version:**  3.9.3

**Search Terms:** narrow object type guard

**Code**

```ts
type DictionaryLike = {
    [index: string]: unknown;
};

type ObjectWithFooStringProp = {
    foo: string;
};

const getObjectWithFooStringProp = (
    input: DictionaryLike,
): ObjectWithFooStringProp | null => {
    if (typeof input.foo === 'string') {
        input.foo; // Type is `string` ✅

        // Unexpected error ❌
        // Property 'foo' is missing in type 'DictionaryLike' but required in type 'ObjectWithFooStringProp'.
        return input;

        // Workaround:
        // return { foo: input.foo };
    } else {
        return null;
    }
};
```

**Expected behavior:**

No errors

**Actual behavior:**

Errors

**Playground Link:** [Link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=22&pc=2#code/C4TwDgpgBAIglgY2HA9gOwIYCcQBk4DW0AvFAN4BQU1UA2nGgCYQAeAXFAM7BYMDmAXQ4BXNATQoA7mgDcFAL5yKoSFADyAIwBWEJAHU4wABYAxFCgDKPfgAUsKMFFKUaUAGbmO3Xmj5zFFBQI6NxQfBDAmjr6hqbmVj58dg5OUAAUVDQMYMLAHPBIqJg4+EQANBQAlBxRusAGxmaW1r7JjgA+UGjCADY9TgB85JnUcG7pKhAo49m5AHQeKE7EpADk3vyrlcOurrPAC+YyUAD0J1AAKuDQcJxQAAYbvvdQgKDkgbs0Z1AAqmiskCQEEYUAgWHsWCggBlyEa7b5tMGgKCrRarKC3KAAW1unH46LQUEmyIKyHQ2DwhAgaI0uSgWAgAEdhHB6SCGITrsjajFGvEWkl7GBVnNYa56cBhFgCfslJ9qN89CgsARsChRIw2KKvudxZKCWR3J58TkDosoAFXPJQT1ONAXHLdVKur0enJLQoZEA)

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
Duplicate #10065

**Comment by OliverJAsh** [comment]
Thanks. From the discussion in that issue it's not clear what the current status of this is. Is it still considered an open suggestion or is it a "won't fix"?
